### PR TITLE
[FIRRTL] Add a parser for `firrtl.instance_choice`

### PR DESCRIPTION
**Pull request by nandor**

### CODE REVIEW

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp [hunk 1]
```diff
@@ -4868,6 +4974,10 @@ ParseResult FIRCircuitParser::parseToplevelDefinition(CircuitOp circuit,
     return emitError(getToken().getLoc(), "only modules may be public");
   case FIRToken::kw_type:
     return parseTypeDecl();
+  case FIRToken::kw_option:
+    if (requireFeature(nextFIRVersion, "options"))
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Import/FIRParser.cpp:4978]: I'd go with the full feature name here or possibly both: 
```suggestion
    if (requireFeature(nextFIRVersion, "option groups/instance choices"))
```

Comment (seldridge) [reply lib/Dialect/FIRRTL/Import/FIRParser.cpp:4978]: Also, new features usually lock in this check in a test.

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp [hunk 2]
```diff
@@ -3785,6 +3771,124 @@ ParseResult FIRStmtParser::parseInstance() {
...
+ParseResult FIRStmtParser::parseInstanceChoice() {
+  auto startTok = consumeToken(FIRToken::kw_instchoice);
+  SMLoc loc = startTok.getLoc();
+
+  // If this was actually the start of a connect or something else handle that.
+  if (auto isExpr = parseExpWithLeadingKeyword(startTok))
+    return *isExpr;
+
+  if (requireFeature(nextFIRVersion, "options"))
+    return failure();
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Import/FIRParser.cpp:3785]: Same comment as elsewhere about the message here and adding tests.

Also, the usage of `nextFIRVersion` is probably imprecise for this. I think instance choices are going to make 4.0.0?

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp [hunk 3]
```diff
@@ -3785,6 +3771,124 @@ ParseResult FIRStmtParser::parseInstance() {
...
+    if (!optionCase)
+      return emitError(loc, "use of undefined option case '" + caseId + "'");
+    caseModules.emplace_back(optionCase, caseModule);
+  }
+
+  auto annotations = getConstants().emptyArrayAttr;
+  SmallVector<Attribute, 4> portAnnotations(modulePorts.size(), annotations);
+
+  // Create an instance choice op.
+  StringAttr sym = {};
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/Import/FIRParser.cpp:3857]: Nit: StringAttr probably doesn't need the empty initializer list? Is the following fine:

```
StringAttr sym;
```

**Review by seldridge — COMMENTED** [review]
Please add emitter support so that that doesn't get out of date with new features.

Also, include tests for the various error cases here (at least as consistently as other features are tested here, e.g., see how layers are tested in `parse-errors.mlir`).

File: test/firtool/instchoice.fir [hunk 4]
```diff
@@ -0,0 +1,32 @@
+; RUN: firtool %s --parse-only | FileCheck %s
+
+FIRRTL version 4.0.0
+circuit Foo:
+  option Platform:
+    case FPGA:
+    case ASIC:
```
[the remark this answers is not in the document]
Comment (nandor) [reply test/firtool/instchoice.fir:7]: I wanted to leave this open for extension, but I now think it's better to drop it.

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp [hunk 5]
```diff
@@ -4904,6 +5014,46 @@ ParseResult FIRCircuitParser::parseTypeDecl() {
...
 
+// Parse an option group declaration.
+ParseResult FIRCircuitParser::parseOptionDecl(CircuitOp circuit) {
+  StringRef id;
+  consumeToken();
+  auto loc = getToken().getLoc();
+
+  if (getToken().isKeyword())
+    return emitError(loc) << "cannot use keyword '" << getToken().getSpelling()
+                          << "' for option group name";
```
[the remark this answers is not in the document]
Comment (nandor) [reply lib/Dialect/FIRRTL/Import/FIRParser.cpp:5025]: 🙄

File: test/firtool/instchoice.fir [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    input clock: Clock
+
+  module FPGATarget:
+    input clock: Clock
+
+  module ASICTarget:
+    input clock: Clock
+
+  module Foo:
+    input clock: Clock
```
[the remark this answers is not in the document]
Comment (nandor) [reply test/firtool/instchoice.fir:19]: I disagree, the resulting `firrtl.instchoice` op does have ports that are creating by the parser.

File: test/firtool/instchoice.fir [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+    ; CHECK: %inst_clock = firrtl.instance_choice inst interesting_name @DefaultTarget 
+    ; CHECK-SAME: alternatives @Platform { 
+    ; CHECK-SAME:   @FPGA -> @FPGATarget, 
+    ; CHECK-SAME:   @ASIC -> @ASICTarget 
+    ; CHECK-SAME: } (in clock: !firrtl.clock)
+    instchoice inst of DefaultTarget Platform
+      FPGA => FPGATarget
+      ASIC => ASICTarget
```
[the remark this answers is not in the document]
Comment (nandor) [reply test/firtool/instchoice.fir:28]: Commas seem to be treated as whitespace?

**Comment by nandor** [comment]
Pulled the emitter out to #6520

File: test/firtool/instchoice.fir (hunk 7 above)
[the remark this answers is not in the document]
Comment (seldridge) [reply test/firtool/instchoice.fir]: Good eye... commas are historically treated as whitespace. However, the spec doesn't think this is how things work. Keep the comma in the test as I want to make commas mandatory in the future.

File: test/firtool/instchoice.fir [hunk 8]
```diff
@@ -0,0 +1,32 @@
+; RUN: firtool %s --parse-only | FileCheck %s
+
+FIRRTL version 4.0.0
+circuit Foo:
+  option Platform:
+    FPGA
+    ASIC
```
Comment (seldridge) [inline test/firtool/instchoice.fir:7]: Interesting. Yes, the `case` keyword was probably adding nothing here. I'm good with this.

File: test/firtool/instchoice.fir [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    input clock: Clock
+
+    ; CHECK: %inst_clock = firrtl.instance_choice inst interesting_name @DefaultTarget 
+    ; CHECK-SAME: alternatives @Platform { 
+    ; CHECK-SAME:   @FPGA -> @FPGATarget, 
+    ; CHECK-SAME:   @ASIC -> @ASICTarget 
+    ; CHECK-SAME: } (in clock: !firrtl.clock)
+    instchoice inst of DefaultTarget, Platform
+      FPGA => FPGATarget
+      ASIC => ASICTarget
```
Comment (seldridge) [inline test/firtool/instchoice.fir:28]: This really should have a trailing `:` to be consistent with other operations which create new scopes.

File: test/firtool/instchoice.fir (hunk 8 above)
Comment (seldridge) [inline test/firtool/instchoice.fir:7]: Please add checking that the options themselves are parsed correctly. 🥺

File: test/Dialect/FIRRTL/parse-errors.fir [hunk 10]
```diff
@@ -1154,3 +1154,95 @@ circuit RWProbeExprOOB :
...
+      ASIC => FPGATarget
+
+;// -----
+
+FIRRTL version 4.0.0
+circuit Foo:
+  ; expected-error @below {{duplicate option case definition 'FPGA'}}
+  option Platform:
+    FPGA
+    FPGA
```
Comment (seldridge) [inline test/Dialect/FIRRTL/parse-errors.fir:1248]: Great looking tests, now. 💯

**Review by seldridge — APPROVED** [review]
LGTM

**Pull request merged**